design/bridge_design.md

```diff
@@ -0,0 +1,317 @@
+
+```solidity 
+function submitOutOfNetworkMessage(
+        address      sender,
+        uint??       sequence,
+        uint32       nonce,
+        bytes memory topic,
+        bytes memory payload
+) public onlyRole(ADMINISTRATOR|OWNER)
+```
+
+This function should not be callable by any users or contracts unless they have the `ADMINISTRATOR` role or are the `Owner` of the `MessageBus` contract. Along with those requirements, on the Obscuro layer 2, there is an additional security measure - Any incoming transaction to this function is blocked, even if it is coming from an address that has the correct role. The protocol will block transactions that do not originate from inside of the enclave due to consuming L1 blocks. This means that even if the enclave gets hacked and the keys leak the verifiers will block withdraws.
+
+When called, the function should store the message indexed in storage - map of `senders` that contains a map of `topics` which points to array of `messages` should be sufficient. 
+Along with it, the hash of the whole message should be used as a key to store a message received flag inside the map `receivedMessages`. This map enables quick verification that a message is valid and received.
+
+### Query Messages
+
+This function allows querying messages on a topic, by a certain user. 
+
+```solidity 
+function queryMessages(
+        address      sender,
+        bytes memory topic,
+        uint256      fromIndex,
+        uint256      toIndex,
+) public returns (bytes [] memory)
+```
+
+This function should return all the messages between the requested `fromIndex` and `toIndex` of the array storing the submitted messages. The array can be found when going through the map of `senders` under the concrete `topic`. It is a more expensive way to get messages on-chain without relying on users to ferry them around. 
+
+### MessageBus Internal workings - L1 to L2
+
+When a block from `L1` is processed by the `enclave` and transactions inside of it result in `MessagePublished` events, the `enclave` will submit them as messages to the inbox of the `L2` contract. This enables the `L2 smart contracts` to use/consume messages from `L1`. 
+
+![Diagram not found](./resources/PublishFromEthereum.svg)
+
+> **_NOTE:_** The enclave should not accept user transactions to system contracts. Only internal/synthetic transactions should be able to call the system functions!
+
+### MessageBus Internal workings - L2 to L1
+
+When a transaction on the `L2` results in `LogMessagePublished`, the event will automatically be added to the `Rollup header` by the `enclave`. Then the management contract will submit them to the `MessageBus` or they will directly be.
+
+![Diagram not found](./resources/PublishFromObscuro.svg)
```

Comment:
a cheaper alternative is to publish the root hash of the event, and make it the user's responsibillity to claim the results using the proof

Reply:
I was thinking of this, but in a slightly different manner where the user can publish the full message if it matches the published hash. I'm quite undecided on the best approach at this point. I guess we can discuss the options in the design review meeting and agree on a direction for phase 1.

Reply:
would be good to list all the options here